src/service/viewer-impl.js

```diff
@@ -1120,30 +1135,31 @@ export class Viewer {
   }
 
   /**
-   * Replace the document url with the viewer provided new replaceUrl.
-   * @param {?string} newUrl
+   * Attempts to generate a replace URL using the "amp_r" query parameter,
+   * returning null if invalid.
+   * @return {?string}
    */
-  replaceUrl(newUrl) {
-    if (!newUrl ||
-        !this.ampdoc.isSingleDoc() ||
-        !this.win.history.replaceState) {
-      return;
+  generateReplaceUrl_() {
```

Comment:
I think this should belong to docucment-info-impl.js

Reply:
I thought I'd put it here so that it can be computed once when the viewer is loaded, and then retrieved later. But if it'll only be needed once, perhaps it's best to put it in doc-info? What do you think is best?